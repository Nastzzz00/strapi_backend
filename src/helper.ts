const { createApolloFetch } = require('apollo-fetch')
import { execute, makePromise } from 'apollo-link'
import { HttpLink } from 'apollo-link-http'
import gql from 'graphql-tag'
const { EPOD_API_URI } = process.env

require('dotenv').config()

export const fetchProducts = async (header) => {
  const uri = EPOD_API_URI || 'http://localhost:1337/playground'
  const link = new HttpLink({ uri })
  const operation = {
    query: gql`
      query {
        products {
          id
          created_at
          updated_at
          name
          description
          amount
          productCode
          quantity
          Image {
            id
            created_at
            updated_at
            name
            caption
            alternativeText
            width
            height
            formats
            hash
            ext
            mime
            size
            url
            provider
            previewUrl
            provider_metadata
            related {
              __typename
            }
          }
        }
      }
    `,
    variables: {},
    context: {
      headers: {
        Authorization: header,
      },
    },
  }

  const result: any = await makePromise(execute(link, operation))
    .then((data) => data)
    .catch((error) => error)

  return result.data || result.error
}

export const createProduct = (header) => async (product) => {
  const uri = EPOD_API_URI || 'http://localhost:1337/playground'

  const link = new HttpLink({ uri })

  const mutateProduct = gql`
    mutation a($product: createProductInput) {
      createProduct(product: $product) {
        id
        created_at
        updated_at
        name
        description
        amount
        productCode
        quantity
        Image {
          id
          created_at
          updated_at
          name
          caption
          alternativeText
          width
          height
          formats
          hash
          ext
          mime
          size
          url
          provider
          previewUrl
          provider_metadata
          related {
            __typename
          }
        }
      }
    }
  `

  const operation = {
    query: mutateProduct,
    variables: { product: product },
    context: {
      headers: {
        //Den auth
        Authorization: header,
        //Mark auth
        //Authorization: "Basic bWFyazoxMjM=",
      },
    },
  }

  const result: any = await makePromise(execute(link, operation))
    .then((data) => data)
    .catch((error) => error)
  return result.data || result.error
}
