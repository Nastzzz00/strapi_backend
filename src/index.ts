import { ApolloServer } from "apollo-server";

import { resolvers, typeDefs } from "./graphql";

const server = new ApolloServer({
  typeDefs,
  resolvers,
});

// You can also easily pass variables for dynamic arguments
