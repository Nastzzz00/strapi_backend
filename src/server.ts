require('dotenv').config()
import { ApolloServer } from 'apollo-server-express'
import express from 'express'
import { resolvers, typeDefs } from './graphql'

import {
  fetchProducts,
  fetchOneProduct,
  createProduct,
  updateProduct,
  deleteProduct,
  fetchOrders,
  fetchOneOrder,
  createOrder,
  updateOrder,
  deleteOrder,
} from './helper'

const { authURI, PORT, EPOD_API_URI } = process.env

const startServer = async () => {
  const server = await new ApolloServer({
    typeDefs,
    resolvers,
    context: async (session) => ({
      fetchProducts: async () =>
        fetchProducts(session.req.headers.authorization),
      fetchOneProduct: async () =>
        fetchOneProduct(session.req.headers.authorization),
      createProduct: async () =>
        createProduct(session.req.headers.authorization),
      updateProduct: async () =>
        updateProduct(session.req.headers.authorization),
      deleteProduct: async () =>
        deleteProduct(session.req.headers.authorization),
      fetchOrders: async () => fetchOrders(session.req.headers.authorization),
      fetchOneOrder: async () =>
        fetchOneOrder(session.req.headers.authorization),
      createOrder: async () => createOrder(session.req.headers.authorization),
      updateOrder: async () => updateOrder(session.req.headers.authorization),
      deleteOrder: async () => deleteOrder(session.req.headers.authorization),
    }),
  })
  const app = express()
  const port = PORT || 5003
  server.applyMiddleware({ app })

  app.listen({ port: port }, () =>
    console.log(
      `ðŸš€  Server ready at http://localhost:${port}${server.graphqlPath}`,
    ),
  )
}

startServer()
