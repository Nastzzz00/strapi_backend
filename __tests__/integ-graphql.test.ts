import { createTestClient } from 'apollo-server-testing'
import gql from 'graphql-tag'
import { constructTestServer } from './__utils'
import { createProduct } from '../src/helper'

const { server }: any = constructTestServer({
  context: {
    createProduct: jest.fn(() => async (product) => {
      return {
        data: {
          createProduct: {
            product: {
              id: '1',
              ...product.data,
            },
          },
        },
      }
    }),
    deleteProduct: jest.fn(() => async (product) => {
      return {
        data: {
          deleteProduct: {
            product: {
              id: '1',
              ...product.data,
            },
          },
        },
      }
    }),
    updatedProduct: jest.fn(() => async (product) => {
      return {
        data: {
          createProduct: {
            product: {
              id: '1',
              ...product.data,
            },
          },
        },
      }
    }),
  },
})

describe('Queries', () => {
  //Mutations
  it('create a product', async () => {
    const CREATE_PRODUCT = gql`
      mutation a($input: createProductInput) {
        createProduct(input: $input) {
          product {
            id
            created_at
            updated_at
            name
            description
            amount
            productCode
            quantity
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: CREATE_PRODUCT,
      variables: {
        input: {
          data: {
            name: 'HOTDOG',
            description: 'Hotdog numbawan',
            amount: '123',
            productCode: '420',
            quantity: '100',
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  it('update a product', async () => {
    const UPDATE_PRODUCT = gql`
      mutation a($input: createProductInput) {
        createProduct(input: $input) {
          product {
            id
            created_at
            updated_at
            name
            description
            amount
            productCode
            quantity
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: UPDATE_PRODUCT,
      variables: {
        input: {
          data: {
            name: 'HOTDOG',
            description: 'Hotdog numbawan',
            amount: '123',
            productCode: '420',
            quantity: '100',
          },
        },
      },
    })

    expect(res.errors).toBeUndefined()
    expect(res.data).toMatchObject({
      createProduct: {
        product: {
          id: '1',
          name: 'HOTDOG',
          description: 'Hotdog numbawan',
          amount: '123',
          productCode: '420',
          quantity: '100',
        },
      },
    })
    expect(res).toMatchSnapshot()
  })

  it('delete a product', async () => {
    const DELETE_PRODUCT = gql`
      mutation a($input: deleteProductInput) {
        deleteProduct(input: $input) {
          product {
            id
            created_at
            updated_at
            name
            description
            amount
            productCode
            quantity
            Image {
              id
              created_at
              updated_at
              name
              caption
              alternativeText
              width
              height
              hash
              ext
              mime
              size
              url
              provider
              previewUrl
            }
          }
        }
      }
    `

    const { mutate } = createTestClient(server)
    const res = await mutate({
      mutation: DELETE_PRODUCT,
      variables: {
        input: {
          where: {
            id: '1',
          },
        },
      },
    })

    expect(res).toMatchSnapshot()
  })
})
