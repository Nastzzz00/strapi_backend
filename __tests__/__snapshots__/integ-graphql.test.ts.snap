// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries create a product 1`] = `
Object {
  "data": Object {
    "createProduct": Object {
      "product": Object {
        "amount": "123",
        "created_at": null,
        "description": "Hotdog numbawan",
        "id": "1",
        "name": "HOTDOG",
        "productCode": "420",
        "quantity": "100",
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries delete a product 1`] = `
Object {
  "data": Object {
    "deleteProduct": Object {
      "product": Object {
        "Image": null,
        "amount": null,
        "created_at": null,
        "description": null,
        "id": "1",
        "name": null,
        "productCode": null,
        "quantity": null,
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries update a product 1`] = `
Object {
  "data": Object {
    "createProduct": Object {
      "product": Object {
        "amount": "123",
        "created_at": null,
        "description": "Hotdog numbawan",
        "id": "1",
        "name": "HOTDOG",
        "productCode": "420",
        "quantity": "100",
        "updated_at": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
